From 004c09e8b9fc2182176dfea9b96cef2f1446d081 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ohtokani=20Pinz=C3=B3n?=
 <nxf81389@lsvm07u0000029.swis.us-west-2.aws.nxp.com>
Date: Thu, 29 May 2025 02:16:34 -0700
Subject: [PATCH] Fixes in wayland.c
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

-Added window management (wm) capabilities
-Using 2 draw buffers instead of 1
-Removed wait callback

Signed-off-by: Ohtokani Pinzón <nxf81389@lsvm07u0000029.swis.us-west-2.aws.nxp.com>
---
 wayland/wayland.c | 56 ++++++++++++++++++++++++++++++++---------------
 1 file changed, 38 insertions(+), 18 deletions(-)

diff --git a/wayland/wayland.c b/wayland/wayland.c
index acf79ac..82cfe2b 100644
--- a/wayland/wayland.c
+++ b/wayland/wayland.c
@@ -1,3 +1,8 @@
+/*
+ * Copyright 2023 NXP
+ * SPDX-License-Identifier: MIT
+ */
+
 /**
  * @file wayland.c
  *
@@ -122,7 +127,6 @@ struct buffer_hdl
     int size;
     struct wl_buffer *wl_buffer;
     bool busy;
-    lv_disp_drv_t * disp_waiting_flush_ready;
 };
 
 struct buffer_allocator
@@ -221,6 +225,7 @@ struct window
 #if LV_WAYLAND_XDG_SHELL
     struct xdg_surface *xdg_surface;
     struct xdg_toplevel *xdg_toplevel;
+    uint32_t wm_capabilities;
 #endif
 
     struct buffer_allocator allocator;
@@ -1145,10 +1150,23 @@ static void xdg_toplevel_handle_configure_bounds(void *data, struct xdg_toplevel
      */
 }
 
+static void xdg_toplevel_handle_wm_capabilities(void *data, struct xdg_toplevel *xdg_toplevel,
+                                                struct wl_array *capabilities)
+{
+    uint32_t *cap;
+    struct window *window = (struct window *)data;
+
+    wl_array_for_each(cap, capabilities) {
+        window->wm_capabilities |= (1 << (*cap));
+        /* TODO: Disable appropriate graphics/capabilities as appropriate */
+    }
+}
+
 static const struct xdg_toplevel_listener xdg_toplevel_listener = {
     .configure = xdg_toplevel_handle_configure,
     .close = xdg_toplevel_handle_close,
-    .configure_bounds = xdg_toplevel_handle_configure_bounds
+    .configure_bounds = xdg_toplevel_handle_configure_bounds,
+    .wm_capabilities = xdg_toplevel_handle_wm_capabilities
 };
 
 static void xdg_wm_base_ping(void *data, struct xdg_wm_base *xdg_wm_base, uint32_t serial)
@@ -1214,10 +1232,6 @@ static void handle_wl_buffer_release(void *data, struct wl_buffer *wl_buffer)
 {
     struct buffer_hdl *buffer_hdl = (struct buffer_hdl *)data;
     buffer_hdl->busy = false;
-    if(buffer_hdl->disp_waiting_flush_ready) {
-        lv_disp_flush_ready(buffer_hdl->disp_waiting_flush_ready);
-        buffer_hdl->disp_waiting_flush_ready = NULL;
-    }
 }
 
 static const struct wl_buffer_listener wl_buffer_listener = {
@@ -1955,6 +1969,12 @@ static void _lv_wayland_flush(lv_disp_drv_t *disp_drv, const lv_area_t *area, lv
         lv_disp_flush_ready(disp_drv);
         return;
     }
+    else if (buffer->busy)
+    {
+        LV_LOG_WARN("skip flush since wayland backing buffer is busy");
+        lv_disp_flush_ready(disp_drv);
+        return;
+    }
 
     int32_t x;
     int32_t y;
@@ -1993,9 +2013,7 @@ static void _lv_wayland_flush(lv_disp_drv_t *disp_drv, const lv_area_t *area, lv
            wl_surface_commit(window->body->surface);
         }
         buffer->busy = true;
-        buffer->disp_waiting_flush_ready = disp_drv;
         window->flush_pending = true;
-        return;
     }
 
     lv_disp_flush_ready(disp_drv);
@@ -2118,13 +2136,6 @@ static void _lv_wayland_cycle(lv_timer_t * tmr)
     _lv_wayland_handle_output();
 }
 
-static void _lv_wayland_wait(lv_disp_drv_t * disp_drv)
-{
-    LV_UNUSED(disp_drv);
-    _lv_wayland_handle_input();
-    _lv_wayland_handle_output();
-}
-
 static void _lv_wayland_pointer_read(lv_indev_drv_t *drv, lv_indev_data_t *data)
 {
     struct window *window = drv->disp->driver->user_data;
@@ -2323,6 +2334,7 @@ lv_disp_t * lv_wayland_create_window(lv_coord_t hor_res, lv_coord_t ver_res, cha
                                      lv_wayland_display_close_f_t close_cb)
 {
     lv_color_t * buf1 = NULL;
+    lv_color_t * buf2 = NULL;
     struct window *window;
 
     window = create_window(&application, hor_res, ver_res, title);
@@ -2338,12 +2350,20 @@ lv_disp_t * lv_wayland_create_window(lv_coord_t hor_res, lv_coord_t ver_res, cha
     buf1 = lv_mem_alloc(hor_res * ver_res * sizeof(lv_color_t));
     if (!buf1)
     {
-        LV_LOG_ERROR("failed to allocate draw buffer");
+        LV_LOG_ERROR("failed to allocate draw buffer 1");
+        destroy_window(window);
+        return NULL;
+    }
+
+    buf2 = lv_mem_alloc(hor_res * ver_res * sizeof(lv_color_t));
+    if (!buf2)
+    {
+        LV_LOG_ERROR("failed to allocate draw buffer 2");
         destroy_window(window);
         return NULL;
     }
 
-    lv_disp_draw_buf_init(&window->lv_disp_draw_buf, buf1, NULL, hor_res * ver_res);
+    lv_disp_draw_buf_init(&window->lv_disp_draw_buf, buf1, buf2, hor_res * ver_res);
 
     /* Initialize display driver */
     lv_disp_drv_init(&window->lv_disp_drv);
@@ -2351,8 +2371,8 @@ lv_disp_t * lv_wayland_create_window(lv_coord_t hor_res, lv_coord_t ver_res, cha
     window->lv_disp_drv.hor_res = hor_res;
     window->lv_disp_drv.ver_res = ver_res;
     window->lv_disp_drv.flush_cb = _lv_wayland_flush;
-    window->lv_disp_drv.wait_cb = _lv_wayland_wait;
     window->lv_disp_drv.user_data = window;
+	window->lv_disp_drv.full_refresh = 1;
 
     /* Register display */
     window->lv_disp = lv_disp_drv_register(&window->lv_disp_drv);
-- 
2.34.1

